File: 0940-fruit-into-baskets/0940-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:

        b={}
        left=0
        res=0
        for i in range(len(fruits)):
            curr=fruits[i]
            b[curr]=b.get(curr,0)+1
            while(len(b)>2):
                pre=fruits[left]
                b[pre]-=1
                if b[pre]==0:
                    del b[pre]
                left+=1
            res=max(res,i-left+1)
        return res
        

File: 0940-fruit-into-baskets/0940-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int left=0;
        int res=0;
        HashMap<Integer,Integer> b=new HashMap<>();
        for(int i=0;i<fruits.length;i++){
            int t=fruits[i];
            b.put(t,b.getOrDefault(t,0)+1);
            while(b.size()>2){
                int y=fruits[left];
                b.put(y,b.get(y)-1);
                if (b.get(y)==0){
                    b.remove(y);
                }
                left++;
            }  res=Math.max(res,i-left+1);      }return res;
    }
}